### PR TITLE
Replace linear diophantine solver to resolve issue #9539

### PR DESCRIPTION
**Pull request by Gadal**
... and always return a valid solution when a solution exists.

### CODE REVIEW

**Comment by Gadal** [comment]
d152057 needs to be merged.

File: sympy/solvers/diophantine.py [hunk 1]
```diff
@@ -381,45 +381,113 @@ def diop_linear(eq, param=symbols("t", integer=True)):
...
+               g_0       g_0
+
+    Then we can solve for x_0, y_0 with base_solution_linear().
+
+    x_0 is appended to our return value, while y_0 is used to 
+    solve for x_1 and x_2 using base_solution_linear() again.
+    """
+    
+    if len(var) < 2:
+        raise ValueError("Fewer than two elements provided for 'var' at _diop_linear()")
```
Comment (jksuom) [inline sympy/solvers/diophantine.py]: An equation with a single variable may also have a solution. If it doesn't, ``(None,)`` could be returned. (According to the [documentation](http://docs.sympy.org/latest/modules/solvers/diophantine.html#tutorial), in case of no solutions, a tuple of ``None`` is returned.)

Comment (Gadal) [inline sympy/solvers/diophantine.py]: See line 187, where univariate input is handled.

Comment (Gadal) [inline sympy/solvers/diophantine.py]: Nevermind, that doesn't output in the right format ((None), as you mentioned) and should be in diop_solve anyway.

**Comment by Gadal** [comment]
@aktech 

The Travis build is failing on tests from test_diophantine.py that are changed by the pull request.  How do I go about getting Travis to validate based on the new tests instead of the old?

**Comment by aktech** [comment]
> How do I go about getting Travis to validate based on the new tests instead of the old?

Travis doesn't only tests the new tests. It tests all the test in sympy, so if the old tests are obsolete now, you should replace them with new tests, either equivalent to old or better than old.

**Comment by Gadal** [comment]
I thought I had replaced the old tests (in /sympy/solvers/tests/test_diophantine.py), but when the build fails and it prints the result of the test, it's the old ones that are failing.  For example:

```
Failed example:
    diop_linear(2*x - 3*y - 4*z -3)
Expected:
    (-3*t - 4*z - 3, -2*t - 4*z - 3,  z)
Got:
    (t_0, -6*t_0 - 4*t_1 + 3, 5*t_0 + 3*t_1 - 3)
```

Whereas the new tests just validate that the solution satisfies the constraints of the problem:
```
2*(t_0) - 3*(-6*t_0 - 4*t_1 + 3) - 4*(5*t_0 + 3*t_1) 
== 2*t_0 +18*t_0 - 20*t_0 + 12*t_1 - 12*t_1 + 3 
== 3
```

In any case, the changes that I've made to the test suite doesn't appear to be seen by Travis.  Any idea why that might be?

**Comment by aktech** [comment]
The failing test that you have pointed is in the doc strings of:

```python
def diop_linear(eq, param=symbols("t", integer=True))
```
in `diophantine.py` file.
See [this](https://github.com/sympy/sympy/blob/master/sympy/solvers/diophantine.py#L368), it needs to be changed.

File: sympy/solvers/tests/test_diophantine.py [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
+                   factor_list,
+                   igcd,
+                   Integer,
+                   Matrix,
+                   Mul,
+                   S,
+                   simplify,
+                   Subs,
+                   Symbol,
+                   symbols)
```
Comment (asmeurer) [inline sympy/solvers/tests/test_diophantine.py]: Why did you do this? The imports should be formatted as they were.

File: sympy/solvers/tests/test_diophantine.py [hunk 3]
```diff
@@ -583,6 +582,58 @@ def check_solutions(eq):
...
     return okay
 
+
+def check_linear(eq, param=symbols("t", Integer=True)):
+    """
+    Determines whether solutions returned by diop_solve() satisfy the
+    original equation.
+    """
+
+    eq = eq.expand(force=True)
```
Comment (asmeurer) [inline sympy/solvers/tests/test_diophantine.py]: Why force=True?

File: sympy/solvers/tests/test_diophantine.py [hunk 4]
```diff
@@ -583,6 +582,58 @@ def check_solutions(eq):
...
+
+    var = []
+    coeff = {}
+    diop_type = None
+
+    if type(eq) is Add:
```
Comment (asmeurer) [inline sympy/solvers/tests/test_diophantine.py]: Use isinstance

File: sympy/solvers/tests/test_diophantine.py [hunk 5]
```diff
@@ -583,6 +582,58 @@ def check_solutions(eq):
...
+
+    elif type(eq) is Mul:
+        var.append(eq.as_two_terms()[1])
+        coeff = {}
+        coeff[eq.as_two_terms()[1]] = Integer(eq.as_two_terms()[0])
+
+    elif type(eq) is Symbol:
+        var.append(eq)
+        coeff = {}
+        coeff[eq] = Integer(1)
```
Comment (asmeurer) [inline sympy/solvers/tests/test_diophantine.py]: I think this whole thing can be replaced with `eq.as_coefficients_dict()`

File: sympy/solvers/tests/test_diophantine.py [hunk 6]
```diff
@@ -583,6 +582,58 @@ def check_solutions(eq):
 
     return okay
 
+
+def check_linear(eq, param=symbols("t", Integer=True)):
```
Comment (asmeurer) [inline sympy/solvers/tests/test_diophantine.py]: This function seems a bit complex for a test file. Would it be better to put it in the diophantine module itself?

File: sympy/solvers/tests/test_diophantine.py (hunk 2 above)
Comment (Gadal) [inline sympy/solvers/tests/test_diophantine.py]: It took me a moment to tell if various things were already imported, so I made it readable.  I'll revert it.

**Comment by asmeurer** [comment]
@thilinarmtb will you be able to review this?

File: sympy/solvers/tests/test_diophantine.py (hunk 3 above)
Comment (Gadal) [inline sympy/solvers/tests/test_diophantine.py]: I copied the first few lines of classify_diop().  I didn't check whether force was still necessary.

File: sympy/solvers/tests/test_diophantine.py (hunk 6 above)
Comment (Gadal) [inline sympy/solvers/tests/test_diophantine.py]: Just create test_test_diophantine.py :P

I considered doing:

```
_diop_linear(..., test=False)

...

if test:
    if gcd(var) divides c:
        assert c == Add(*[coeff[var[i]]*solutions[i] for i in range(0, len(var))])
    else:
        assert all(solution == None for solution in solutions)
```

Thoughts?

Comment (jksuom) [inline sympy/solvers/tests/test_diophantine.py]: This function seems to check only if the result satisfies the equation. It does not help to verify that it would be a full solution in the way the original tests did.

Comment (Gadal) [inline sympy/solvers/tests/test_diophantine.py]: I disagree.  The original tests only checked that the output matched the test case.  Which meant that when I changed the output's formatting, my valid solutions failed the test cases.  I suspect they were previously written by copy/pasting the output from the function into the test file, which would defeat the purpose of testing.

File: sympy/solvers/tests/test_diophantine.py (hunk 5 above)
Comment (Gadal) [inline sympy/solvers/tests/test_diophantine.py]: Yup.

File: sympy/solvers/tests/test_diophantine.py (hunk 6 above)
Comment (jksuom) [inline sympy/solvers/tests/test_diophantine.py]: The purpose of the test is to catch regressions before they are merged to master.This check does in no way indicate if a solution was reduced to a partial solution after some modification. It will only show that the remaining values do satisfy the equation. 

> The original tests only checked that the output matched the test case. 

I don't think this would necessarily be wrong. A change of output will be quickly revealed.

The result should be verified when the test is written. It should not be necessary to repeatedly verify the correctness of the original result. (The new tests take twice the time of the old test_linear.)

Comment (Gadal) [inline sympy/solvers/tests/test_diophantine.py]: Were there not similar concerns when check_solutions() was added?

Comment (Gadal) [inline sympy/solvers/tests/test_diophantine.py]: The extra time probably comes from the extra tests I added, such as those for univariate linear diophantine equations, because _diop_solve() now handles them (as you suggested it should).

I only see a 7% increase in execution time, comparing apples to apples:

```
    iters = 1000

    start = time.clock()
    for i in range(iters):
        assert diop_solve(x) == (0,)
        assert diop_solve(1*x) == (0,)
        assert diop_solve(3*x) == (0,)
        assert diop_solve(x + 1) == (-1,)
        assert diop_solve(2*x + 1) == (None,)
        assert diop_solve(2*x + 4) == (-2,)
        assert diop_solve(y + x) == (t_0, -t_0)
        assert diop_solve(y + x + 0) == (t_0, -t_0)
        assert diop_solve(y + x - 0) == (t_0, -t_0)
        assert diop_solve(0*x - y - 5) == (-5,)
        assert diop_solve(3*y + 2*x - 5) == (3*t_0 - 5, -2*t_0 + 5)
        assert diop_solve(2*x - 3*y - 5) == (-3*t_0 - 5, -2*t_0 - 5)
        assert diop_solve(-2*x - 3*y - 5) == (-3*t_0 + 5, 2*t_0 - 5)
        assert diop_solve(7*x + 5*y) == (5*t_0, -7*t_0)
        assert diop_solve(2*x + 4*y) == (2*t_0, -t_0)
        assert diop_solve(4*x + 6*y - 4) == (3*t_0 - 2, -2*t_0 + 2)
        assert diop_solve(4*x + 6*y - 3) == (None, None)
        assert diop_solve(0*x + 3*y - 4*z + 5) == (-4*t_0 + 5, -3*t_0 + 5)
        assert diop_solve(4*x + 3*y - 4*z + 5) == (t_0, -4*t_1 + 5, t_0 - 3*t_1 + 5)
        assert diop_solve(4*x + 2*y + 8*z - 5) == (None, None, None)
        assert diop_solve(5*x + 7*y - 2*z - 6) == \
            (t_0, -7*t_0 - 2*t_1 + 6, -22*t_0 - 7*t_1 + 18)
        assert diop_solve(3*x - 6*y + 12*z - 9) == (2*t_0 + 3, t_0 + 2*t_1, t_1)
        assert diop_solve(x + 3*y - 4*z + w - 6) == \
            (t_0, t_0 + t_1, -2*t_0 - 3*t_1 - 4*t_2 - 6, -t_0 - 2*t_1 - 3*t_2 - 6)
        assert diop_solve(2*w + 3*x + 5*y + 7*z + 11*t + 13*X + 17*Y + 19*Z - 23) == \
            (t_0, t_0 + t_1, t_0 + t_1 + t_2, t_0 + t_1 + t_2 + t_3, t_0 + t_1 + t_2 + t_3 + t_4, t_0 + t_1 + t_2 + t_3 + t_4 + t_5, -188*t_0 - 149*t_1 - 98*t_2 - 41*t_3 - 8*t_4 - 2*t_5 + 7*t_6 + 69, 125*t_0 + 99*t_1 + 65*t_2 + 27*t_3 + 5*t_4 + t_5 - 5*t_6 - 46)
        assert diop_solve(2*w + 4*x + 6*y + 8*z + 10*t + 12*X + 14*Y + 16*Z - 1) == \
            (None, None, None, None, None, None, None, None)
    end = time.clock()
    print("first: " + str(end - start))
    
    start = time.clock()
    for i in range(iters):
        assert check_linear(x)
        assert check_linear(1*x)
        assert check_linear(3*x)
        assert check_linear(x + 1)
        assert check_linear(2*x + 1)
        assert check_linear(2*x + 4)
        assert check_linear(y + x)
        assert check_linear(y + x + 0)
        assert check_linear(y + x - 0)
        assert check_linear(0*x - y - 5)
        assert check_linear(3*y + 2*x - 5)
        assert check_linear(2*x - 3*y - 5)
        assert check_linear(-2*x - 3*y - 5)
        assert check_linear(7*x + 5*y)
        assert check_linear(2*x + 4*y)
        assert check_linear(4*x + 6*y - 4)
        assert check_linear(4*x + 6*y - 3)
        assert check_linear(0*x + 3*y - 4*z + 5)
        assert check_linear(4*x + 3*y - 4*z + 5)
        assert check_linear(4*x + 2*y + 8*z - 5)
        assert check_linear(5*x + 7*y - 2*z - 6)
        assert check_linear(3*x - 6*y + 12*z - 9)
        assert check_linear(x + 3*y - 4*z + w - 6)
        assert check_linear(2*w + 3*x + 5*y + 7*z + 11*t + 13*X + 17*Y + 19*Z - 23)
        assert check_linear(2*w + 4*x + 6*y + 8*z + 10*t + 12*X + 14*Y + 16*Z - 1)
    end = time.clock()
    print("second: " + str(end - start))
```

Output:

```
first: 16.363018
second: 17.423285
```

File: sympy/solvers/diophantine.py [hunk 7]
```diff
@@ -232,21 +234,30 @@ def classify_diop(eq):
     ([x, y], {1: 5, x: 1, x**2: 1, y: 0, y**2: 1, x*y: -1}, 'binary_quadratic')
     """
     eq = eq.expand(force=True)
-    var = list(eq.free_symbols)
-    var.sort(key=default_sort_key)
 
-    coeff = {}
+    coeff = coeff = eq.as_coefficients_dict()
```
Comment (thilinarmtb) [inline sympy/solvers/diophantine.py]: Why `coeff = coeff = eq.as_coefficients_dict()` instead of `coeff = eq.as_coefficients_dict()`?

**Comment by thilinarmtb** [comment]
@Gadal: Thank you very much for the fix. +1 to merge as it is.

Did you add a test for the original bug? I couldn't find it.

File: sympy/solvers/diophantine.py (hunk 7 above)
Comment (Gadal) [inline sympy/solvers/diophantine.py]: haaaaaaaa... whoops.  Typo, fixed.

**Comment by thilinarmtb** [comment]
One additional note: Functions `check_linear` and `check_solutions` solutions only check whether
the returned solutions satisfy the original equation. They doesn't check whether the solutions are
complete. We need to look at the number of parameters they return as well.

See the following comment:
https://github.com/sympy/sympy/issues/9539#issuecomment-121571466

**Comment by Gadal** [comment]
Added:

```
# and has infinitely many solutions otherwise.
else:
    # Does the solution satisfy the original equation?
    correct = c == Add(*[coeff[var[i]]*solutions[i] for i in range(0, len(var))])

    # Does the solution have as many free variables as possible?
    free_variable_count = len(set([x for solution in solutions for x in solution.free_symbols]))
    complete = len(solutions) - 1 == free_variable_count

    return (correct and complete) 
```

Thoughts?

**Comment by Gadal** [comment]
It's a shame you can't cancel builds.  You're doomed, 861bfb9!  Don't even try!

**Comment by jksuom** [comment]
Counting the number of parameters does not suffice to prove that the solution is complete. For example, if one parameter, say ``t_0`` is replaced with ``t_0 = 2*t`_0`` (so that ``t_0`` is given even values only), we get an incomplete solution with the correct number of parameters.

**Comment by Gadal** [comment]
I'm not sure how to check that that is not the case.  Suggestions?

**Comment by Gadal** [comment]
Nevermind, you can prove it with Bézout.

**Comment by jksuom** [comment]
I think the only way to check that a result is complete is to compare it to a complete solution. It would probably be best to include such a complete solution in the assertion statement. It would then suffice to prove the completeness of the model solution only once, either by hand or by using a proven algorithm.

**Comment by Gadal** [comment]
> It would then suffice to prove the completeness of the model solution only once, either by hand or by using a proven algorithm.

If you know of a way to prove the completeness of a solution by hand, then I can implement that method in code instead of writing asserts.

If there is a proven algorithm, it is unlikely that its outputs will be identical to ours.  Otherwise, please send me a link to the library.  It's likely that there is documentation of the proof of correctness that I can use to prove the correctness of our output.

**Comment by jksuom** [comment]
The standard theory of linear diophantine equations gives the steps to a complete solution. I have tried to sketch that theory [here](https://gist.github.com/jksuom/695024ec938a1a49d893). It not hard to turn that into an algorithm. The result of that algorithm could be used as a model solution in tests.

**Comment by Gadal** [comment]
My pull request is more or less the algorithm you've just described.  And I have a similar sketch in the comments of the code I've submitted.  Writing the same algorithm twice doesn't verify the output, because both might be wrong in the same way.

I am concerned that we might be talking past one another.

@jksuom @thilinarmtb Would it be reasonable to merge this pull request, and open a new issue along the lines of "Improve test coverage for linear diophantine equation solver"?

**Comment by Gadal** [comment]
I replaced check_linear() with asserts.  I think the tests were the last of the concerns.  Anything else?

**Comment by jksuom** [comment]
Thanks, I think the tests should now be adequate. I find it hard to follow the algorithm, but it seems to produce correct solutions. It also appears that the algorithm is of quadratic complexity with its double loop, though that may not matter much with a reasonable number of variables.

**Comment by Gadal** [comment]
Yeah, it is quadratic.  I've been trying to think of a way to avoid that, but nothing comes to mind.

**Comment by Gadal** [comment]
@asmeurer Is this ready to be pulled?

**Comment by Gadal** [comment]
It's been fifteen days.  Is this not being pulled because it's a bad fix?

**Comment by hargup** [comment]
> It's been fifteen days. Is this not being pulled because it's a bad fix?

I'm sorry, the PR are probably not pulled because the involved people were busy in the meanwhile and people like me don't have enough expertise in this area of code have a say. @thilinarmtb @asmeurer can you review the PR and merge it if it is good.

File: sympy/solvers/diophantine.py [hunk 8]
```diff
@@ -381,45 +392,196 @@ def diop_linear(eq, param=symbols("t", integer=True)):
 
 
 def _diop_linear(var, coeff, param):
```
Comment (hargup) [inline sympy/solvers/diophantine.py]: Why is this a private function? I think the function is useful enough to become the part of the public api.

File: sympy/solvers/diophantine.py [hunk 9]
```diff
@@ -381,45 +392,196 @@ def diop_linear(eq, param=symbols("t", integer=True)):
...
+    """
+    Solves diophantine equations of the form:
```
Comment (hargup) [inline sympy/solvers/diophantine.py]: The function needs to have doc tests (code examples).

File: sympy/solvers/diophantine.py (hunk 8 above)
Comment (hargup) [inline sympy/solvers/diophantine.py]: sorry didn't notice `diop_linear`.

**Comment by hargup** [comment]
@Gadal I read the code and it looks good to me, your well written docstrings and @jksuom 's [gist](https://gist.github.com/jksuom/695024ec938a1a49d893) were very helpful. +1

**Comment by thilinarmtb** [comment]
Hi @Gadal : Sorry for not being able to help you enough with your PR. This is no way a bad fix. You cleaned up the code base quite a bit. I was caught up with my own work and couldn't find time to contribute to either SymPy or SymEngine.

Thank you very much for your work. Looking forward to working with you in future.
